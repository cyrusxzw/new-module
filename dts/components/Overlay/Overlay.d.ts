/// <reference types="react" />
import type { OverlayProps } from './Overlay.types';
declare const Overlay: ({
  className,
  isVisible,
  onClose,
}: OverlayProps) => JSX.Element;
export { Overlay };
