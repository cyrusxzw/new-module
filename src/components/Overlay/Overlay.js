import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { useEscapeKeyListener } from '~/customHooks';
import { Transition } from '~/components/Transition';
import styles from './Overlay.module.css';

const Overlay = ({ className, isVisible, onClose }) => {
  useEscapeKeyListener(onClose);

  const classSet = cx(
    styles.base,
    {
      [styles.isVisible]: isVisible,
    },
    className,
  );

  return (
    <Transition
      isActive={!!isVisible}
      shouldMountOnEnter={true}
      shouldUnmountOnExit={true}
      type="fade"
    >
      <div
        aria-hidden="true"
        className={classSet}
        data-testid="data-testid-Overlay"
        onClick={onClose}
        role="button"
        tabIndex={-1}
      />
    </Transition>
  );
};

Overlay.propTypes = {
  className: PropTypes.string,
  isVisible: PropTypes.bool,
  onClose: PropTypes.func.isRequired,
};

Overlay.defaultProps = {
  className: undefined,
  isVisible: false,
  onClose: undefined,
};

export { Overlay };
