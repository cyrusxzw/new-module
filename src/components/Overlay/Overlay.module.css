@import 'styles/variables/index';

.base {
  position: fixed;
  z-index: var(--z-index-overlay);
  top: 0;
  right: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: var(--color-black-rgba-50);
  pointer-events: none;

  &.isVisible {
    pointer-events: auto;
  }
}
