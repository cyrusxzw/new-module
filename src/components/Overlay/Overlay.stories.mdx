import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks';
import * as knobs from '@storybook/addon-knobs';
import { State } from '~.storybook/State';
import { Button } from '~/components/Button';
import { Overlay } from './Overlay';

<Meta title="Components/Overlay" component={Overlay} />

# Overlay

<Canvas>
  <Story name="Base component">
    <State initialState={{ isVisible: false }}>
      {(state, setState) => {
        return (
          <React.Fragment>
            <Overlay
              isVisible={state.isVisible}
              onClose={() => {
                setState(() => ({ ...state, isVisible: false }));
              }}
            />
            <Button
              onClick={() => {
                setState(() => ({ ...state, isVisible: true }));
              }}
              title="Open Overlay"
            >
              Show Overlay
            </Button>
          </React.Fragment>
        );
      }}
    </State>
  </Story>
</Canvas>

## API

<ArgsTable of={Overlay} />
