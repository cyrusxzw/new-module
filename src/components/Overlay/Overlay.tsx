import React from 'react';
import cx from 'classnames';
import { useEscapeKeyListener } from '~/customHooks';
import { Transition } from '~/components/Transition';
import type { OverlayType } from './Overlay.types';
import styles from './Overlay.module.css';

const Overlay: OverlayType = ({ className, isVisible, onClose }) => {
  useEscapeKeyListener(onClose);

  const classSet = cx(
    styles.base,
    {
      [styles.isVisible]: isVisible,
    },
    className,
  );

  return (
    <Transition
      isActive={!!isVisible}
      shouldMountOnEnter={true}
      shouldUnmountOnExit={true}
      type="fade"
    >
      <div
        aria-hidden="true"
        className={classSet}
        data-testid="data-testid-Overlay"
        onClick={onClose}
        role="button"
        tabIndex={-1}
      />
    </Transition>
  );
};

export { Overlay };
