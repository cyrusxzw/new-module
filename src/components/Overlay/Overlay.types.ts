type OverlayProps = {
  className?: string;
  isVisible?: boolean;
  onClose: () => void;
};

export type { OverlayProps };
