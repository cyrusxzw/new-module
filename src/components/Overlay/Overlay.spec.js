import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Overlay } from './Overlay';

describe('<Overlay />', () => {
  const handleOnClose = jest.fn();

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render base component correctly', () => {
    const { container } = render(
      <Overlay isVisible={true} onClose={handleOnClose} />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should call the onClose callback when the escape key is pressed', () => {
    render(<Overlay isVisible={true} onClose={handleOnClose} />);

    const overlay = screen.getByTestId(/data-testid-Overlay/i);

    expect(handleOnClose).not.toHaveBeenCalled();

    userEvent.type(overlay, '{esc}');

    expect(handleOnClose).toHaveBeenCalled();
  });

  it('should call the onClose callback when the overlay is clicked', () => {
    render(<Overlay isVisible={true} onClose={handleOnClose} />);

    const overlay = screen.getByTestId(/data-testid-Overlay/i);

    expect(handleOnClose).not.toHaveBeenCalled();

    userEvent.click(overlay);

    expect(handleOnClose).toHaveBeenCalledTimes(1);
  });
});
