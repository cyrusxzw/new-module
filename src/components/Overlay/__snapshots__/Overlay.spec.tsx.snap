// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Overlay /> should render base component correctly 1`] = `
<div>
  <div
    aria-hidden="true"
    class="base isVisible"
    data-testid="data-testid-Overlay"
    role="button"
    tabindex="-1"
  />
</div>
`;
